### PR TITLE
Add all_reduce_coalesced to functional collectives

### PR DESCRIPTION
**Pull request by kumpera**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98643
* #98642
* #99765
* #99764
* #99763
* __->__ #98640

This adds all_reduce_coalesced to MTPG to ease testing.

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/testing/_internal/distributed/multi_threaded_pg.py [hunk 1]
```diff
@@ -82,12 +82,14 @@ def __init__(self, op):
...
-        for src_rank in range(0, len(data)):
-            tensors.append(data[src_rank][0])
-        res = _reduce_ops[self.op](tensors)
-        for src_rank in range(len(data)):
-            data[src_rank][0].copy_(res)
+        for i in range(len(data[0])):
+            tensors = []
+            for src_rank in range(0, len(data)):
+                tensors.append(data[src_rank][i])
+            res = _reduce_ops[self.op](tensors)
```
Comment (wanchaol) [inline torch/testing/_internal/distributed/multi_threaded_pg.py:89]: does this really work? i.e. `reduce_ops[self.op]` is `torch.sum`, this sum only apply to one tensor and can't append to all the tensors in the coalesing case right?

File: torch/distributed/_functional_collectives.py [hunk 2]
```diff
@@ -375,6 +408,41 @@ def reduce_scatter_tensor(
...
+
+    res = []
+    lookup_tensor = tensor_list[0]
+    wh = WaitHolder(lookup_tensor)
+    for tensor in tensor_list:
+        act = AsyncCollectiveTensor(tensor, wh)
+        res.append(cast(torch.Tensor, act))
+
+    # FIXME we should register all tensors and ref count when to emit the wait
+    _register_wrapper_tensor(res[0], lookup_tensor)
```
Comment (wanchaol) [inline torch/distributed/_functional_collectives.py:443]: shall we fix this inside this PR? i.e. maybe we should make the `AsyncCollectiveTensor` or `aten.wait` works on either a `tensor` or `list_of_tensors`, so that we only need to emit one `wait` call and that `wait` call should wait all tensors inside the list

File: torch/_inductor/ir.py [hunk 3]
```diff
@@ -4113,6 +4113,123 @@ def codegen(self, wrapper):
...
+        group_size: int,
+    ):
+        res = []
+
+        def wrap_input(var):
+            nonlocal res
+            op = ForceInPlace(
+                FlexibleLayout(var.get_device(), var.get_dtype(), var.get_size()), var
+            )
+            res.append(op)
```
Comment (wanchaol) [inline torch/_inductor/ir.py:4176]: should `res` append a `ExternKernel` or a `TensorBox`? I think it should be `TensorBox` as that would be later passed into `AllreduceCoaleseced` inductor node?

File: torch/_inductor/ir.py [hunk 4]
```diff
@@ -4113,6 +4113,123 @@ def codegen(self, wrapper):
...
+        return True
+
+
+class AllReduceCoalesced(ExternKernel):
+    def __init__(self, layout, inputs, constant_args, reduce_op):
+        super().__init__(None, layout, inputs, constant_args)
+        self.reduce_op = reduce_op
+        self.name = V.graph.register_buffer(self)
+
+    def should_allocate(self):
```
Comment (wanchaol) [inline torch/_inductor/ir.py:4157]: nit: similar here, could you document a bit why `should_allocate` returns False here?

File: torch/_inductor/ir.py [hunk 5]
```diff
@@ -4113,6 +4113,123 @@ def codegen(self, wrapper):
         wrapper.writeline(f"_register_tensor_work({output_name}, {output_name}_work)")
 
 
+class MultiOutputNoSizeAssert(MultiOutput):
+    """
+    Extract partial output from a multi-output OP.
+        Works like MultiOutput but doesn't assert size. This must be a property guaranteed by the op emiting this.
```
Comment (wanchaol) [inline torch/_inductor/ir.py:4119]: curious why we don't need size asserts in collectives for the multi output cases?

File: torch/_inductor/ir.py [hunk 6]
```diff
@@ -4113,6 +4113,123 @@ def codegen(self, wrapper):
...
+        output_name = self.get_name()
+        if not wrapper.did_reuse(self, self.inputs[0]):
+            wrapper.writeline(f"{output_name}.copy_({input_name}) #no reuse")
+
+    def __init__(self, layout, input):
+        input = self.realize_input(input)
+        super().__init__(None, layout, self.unwrap_storage([input]), ())
+        self.name = V.graph.register_buffer(self)
+
+    def should_allocate(self):
```
Comment (wanchaol) [inline torch/_inductor/ir.py:4147]: nit: could you add some comment to explain what this function mean and why we need it here?

File: torch/_inductor/ir.py [hunk 7]
```diff
@@ -4113,6 +4113,123 @@ def codegen(self, wrapper):
...
+    Wrap the input of your inplace op to enable this behavior.
+
+    TODO: We should test whether wait_tensor can be a victim of reordering and lead to data races.
+    """
+
+    def codegen(self, wrapper):
+        input_name = self.inputs[0].codegen_reference()
+        output_name = self.get_name()
+        if not wrapper.did_reuse(self, self.inputs[0]):
+            wrapper.writeline(f"{output_name}.copy_({input_name}) #no reuse")
```
Comment (wanchaol) [inline torch/_inductor/ir.py:4140]: testing my understanding, if it's the `reuse` case, we don't need to do anything as inductor would directly take that variable reference?

I thought `ForceInplace` should actually try to reuse the original variable even if `not wrapper.did_reuse`  returns True?

File: torch/distributed/_functional_collectives.py [hunk 8]
```diff
@@ -113,6 +113,19 @@ def _wait_tensor(tensor: torch.Tensor) -> torch.Tensor:
         _wait_and_clear_tensor(data_ptr, version_and_work[0])
     return tensor
 
+class WaitHolder:
+    def __init__(self, lookup_tensor):
```
Comment (wanchaol) [inline torch/distributed/_functional_collectives.py:117]: nit: can you add some documentation too why we need this `WaitHolder`? 

Also, would this `WaitHolder` handle tensor lists directly?

**Review by wanchaol — COMMENTED** [review]
Have some questions and suggestions inlined.

**Review by wanchaol — COMMENTED** [review]
Have some questions and suggestions inlined.

File: test/distributed/test_functional_api.py [hunk 9]
```diff
@@ -225,6 +225,14 @@ def test_all_reduce_eager(self):
         res2 = ft_c.all_reduce(tensor, "sum", (mesh, 1))
         self.assertEqual(res2, torch.tensor([2, 2, 2, 2], dtype=torch.float))
 
+    def test_all_reduce_coalesced_eager(self):
+        tensor = torch.ones([4], device="cpu")
+        mesh = dt.DeviceMesh("cpu", torch.arange(4))
+
+        res = ft_c.all_reduce_coalesced([tensor], "sum", mesh)
```
[the remark this answers is not in the document]
Comment (kumpera) [reply test/distributed/test_functional_api.py:232]: good point.

File: torch/testing/_internal/distributed/multi_threaded_pg.py [hunk 10]
```diff
@@ -289,6 +291,12 @@ def allreduce(self, tensor_list, opts=AllreduceOptions()):
         ProcessLocalGroup._end_coll(coll, self)
         return res
 
+    def allreduce_coalesced(self, tensor_list, opts=AllreduceOptions()):
+        coll = ProcessLocalGroup._start_coll(AllReduce(opts.reduceOp), self)
+        res = coll.join(self._rank, tensor_list)
+        ProcessLocalGroup._end_coll(coll, self)
+        return res
```
[the remark this answers is not in the document]
Comment (kumpera) [reply torch/testing/_internal/distributed/multi_threaded_pg.py:298]: done.

File: torch/testing/_internal/distributed/multi_threaded_pg.py (hunk 1 above)
Comment (kumpera) [reply torch/testing/_internal/distributed/multi_threaded_pg.py:89]: This code is super confusing, it probably deserves some rewriting as it's like 3 levels of nested lists.

We're reducing all tensors in `data[*][i]`. The first dimension is rank.
We reduce input tensors separately.

File: torch/_inductor/ir.py (hunk 3 above)
Comment (kumpera) [reply torch/_inductor/ir.py:4176]: This is really tricky actually
The convention for lowering.py is to return IR nodes and not TensorBox from the `create` funs.
OTOH `AllreduceCoalesced` actually needs a list of `TensorBox` nodes.

This is cleared up in the next PR as I properly decouple inplace in/out nodes from the actual collectives. It took me a dozen iterations to arrive at that design and it's kinda ok.

File: torch/_inductor/ir.py (hunk 5 above)
Comment (kumpera) [reply torch/_inductor/ir.py:4119]: Because we know that the expected size matches the runtime one since we control the allocation.

File: torch/distributed/_functional_collectives.py (hunk 8 above)
Comment (kumpera) [reply torch/distributed/_functional_collectives.py:117]: WaitHolder only handle one tensor. The next PR adds proper handling of tensor lists.

Let me document it though.

**Comment by kumpera** [comment]
Addressed half of the issues, the remained I'll do tomorrow.

File: torch/_inductor/ir.py (hunk 4 above)
Comment (kumpera) [reply torch/_inductor/ir.py:4157]: documented the design at the top of ForceInPlace  including should_allocate.

File: torch/_inductor/ir.py (hunk 7 above)
Comment (kumpera) [reply torch/_inductor/ir.py:4140]: ForceInPlace is a hint to the scheduler (the next PR renamed it to InPlaceHint).

``did_reuse`` here returns whether the input was reused by other, unrelated, ops.

For example:

```python
def kernel_no_reuse(inp):
    buf0 = empty_like(inp) #no reuse
    triton_lalala.run(buf0)

    buf1 = buf0 # reuse it
    dist.all_reduce(buf1)

   buf2 = empty_like(inp)
   triton_lalala.run(buf2)
    return buf1, buf2
```


```python
def kernel_with_reuse(inp):
    buf0 = empty_like(inp) #no reuse
    triton_lalala.run(buf0)

    buf1 = buf0 # reuse it
    dist.all_reduce(buf1)

   buf2 = buf0 #OH NO the scheduler is reusing buf0 instead of allocating a new buffer
   triton_lalala.run(buf2)
    return buf1, buf2
```

The above code will produce the wrong result, so the scheduler plus that codegen check will produce the following instead:


```python
def kernel_with_reuse(inp):
    buf0 = empty_like(inp) #no reuse
    triton_lalala.run(buf0)

    buf1 = empty_like(buf0) # no reuse :( 
    buf1.copy_(buf0)
    dist.all_reduce(buf1)

   buf2 = buf0 #OH NO the scheduler is reusing buf0 instead of allocating a new buffer
   triton_lalala.run(buf2)
    return buf1, buf2
```

This will now produce the right output. The Inductor code is not obvious since allocation happens in the scheduler and is kinda invisible to IR nodes and codegen.

File: torch/distributed/_functional_collectives.py (hunk 8 above)
Comment (kumpera) [reply torch/distributed/_functional_collectives.py:117]: WaitHolder makes a bit more sense in the the PR once we introduce ref counting to wait.

**Review by wanchaol — COMMENTED** [review]
LGTM, I have one more question about whether we should reuse `AsyncCollectiveTensor` by merging the WaitHolder logic into it.

Edit: looks like the above PR removes it

**Review by wanchaol — APPROVED** [review]
one more suggestion if we can add some test to test the inductor codegen.

File: torch/_inductor/ir.py [hunk 11]
```diff
@@ -4113,6 +4113,127 @@ def codegen(self, wrapper):
...
+    - The scheduler special-case this node and enable it to reuse its input.
+    """
+
+    def codegen(self, wrapper):
+        input_name = self.inputs[0].codegen_reference()
+        output_name = self.get_name()
+        if not wrapper.did_reuse(self, self.inputs[0]):
+            wrapper.writeline(f"{output_name}.copy_({input_name}) #no reuse")
+
+    def __init__(self, layout, input):
```
Comment (fegin) [inline torch/_inductor/ir.py:4146]: Can we put `__init__` before `codegen`? The current order is less intuitive.

File: torch/distributed/_functional_collectives.py [hunk 12]
```diff
@@ -141,15 +158,23 @@ def __new__(cls, elem: torch.Tensor):
...
+        self._holder.wait_tensor()
+        return self
+
     @classmethod
     def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
-        def unwrap(e: Any):
-            return wait_tensor(e.elem)
+        def unwrap(e: AsyncCollectiveTensor):
+            # TODO do we need to insert a wait_tensor op for all tensors or just the first one?
+            # Given only first usage of any of the output tensors needs to wait, for now we emit only one wait
```
Comment (fegin) [inline torch/distributed/_functional_collectives.py:175]: Is this the reason why`WaitHolder` is introduced? This is a contract to any graph transformation involving collective. So it may be good to well-document this design.

File: torch/distributed/_functional_collectives.py [hunk 13]
```diff
@@ -113,6 +113,22 @@ def _wait_tensor(tensor: torch.Tensor) -> torch.Tensor:
         _wait_and_clear_tensor(data_ptr, version_and_work[0])
     return tensor
 
+class WaitHolder:
+    """
+    WaitHolder is an indirection to the tensor that needs to be waited on.
```
[the remark this answers is not in the document]
Comment (fegin) [reply torch/distributed/_functional_collectives.py:118]: I'm also confused about why we need a separate class from `AsyncCollectiveTensor`.

File: torch/distributed/_functional_collectives.py [hunk 14]
```diff
@@ -375,6 +411,41 @@ def reduce_scatter_tensor(
     res = _maybe_wrap_tensor(tensor)
     return res
 
+
+def all_reduce_coalesced(self: List[torch.Tensor], reduceOp: str, group: RANK_TYPES, tag: str = "") -> List[torch.Tensor]:
+    """
+    Reduces a list of tensors across all machines in such a way that all get
+    the final result.
+
+    The all tensors in the input list are left unmodified.
```
Comment (fegin) [inline torch/distributed/_functional_collectives.py:420]: Should we mention how wait is performed? That is an important contract.

File: torch/distributed/_functional_collectives.py (hunk 12 above)
Comment (kumpera) [reply torch/distributed/_functional_collectives.py:175]: The follow up PR fixes this to relax this contract and make it more intuitive. All result tensors are waitable.

File: torch/distributed/_functional_collectives.py (hunk 14 above)
Comment (kumpera) [reply torch/distributed/_functional_collectives.py:420]: WaitHolder turned out to be not what we need. I fix it in the next PR
